### PR TITLE
business central advance settings model and service construction

### PR DESCRIPTION
**Pull request by DhaaraniCIT**
## Summary by CodeRabbit

- **New Features**
  - Introduced advanced settings for business operations, including validation rules and configuration options.
  - Added the ability to fetch and update advanced settings related to business central module.
  - Implemented new methods for managing expense fields, admin emails, and expense filters.

- **Enhancements**
  - Improved data handling with new caching strategies for faster and more efficient service responses.

- **Documentation**
  - Updated service documentation to reflect new features and methods available.

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
@coderabbitai review

File: src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts [hunk 1]
```diff
@@ -1,9 +1,58 @@
...
+  }
+
+  @CacheBuster({
+    cacheBusterNotifier: businessCentralAdvancedSettingGetCache$
+  })
+  postAdvancedSettings(advancedSettingsPayload: BusinessCentralAdvancedSettingsPost): Observable<BusinessCentralAdvancedSettingsGet> {
+    return this.apiService.post(`/workspaces/${this.workspaceService.getWorkspaceId()}/advanced_settings/`, advancedSettingsPayload);
+  }
+
+  getExpenseFilelds(): Observable<ConditionField[]> {
```
[the remark this answers is not in the document]
Comment (ashwin1111) [reply src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts]: +1

File: src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import { FormControl, FormGroup } from "@angular/forms";
+import { ConditionField, ExpenseFilterPayload, ExpenseFilterPost } from "../../common/advanced-settings.model";
+import { JoinOption, Operator } from "../../enum/enum.model";
+
+export type AdvancedSettingValidatorRule = {
+    condition1: string[];
+    condition2: string[];
+    operator1: string[];
+    operator2: string[];
+};
```
Comment (ashwin1111) [inline src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts:10]: let's re-use existing type, move existing ones to common folder if it's present in specific app

File: src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    created_at: Date,
+    updated_at: Date,
+    workspace: number
+}
+
+export interface BusinessCentralAdvancedSettingsPost extends BusinessCentralAdvancedSettings { }
+
+export class BusinessCentralAdvancedSettingsModel {
+
+    static constructExportFilterPayload(valueField: any): ExpenseFilterPayload {
```
Comment (ashwin1111) [inline src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts:30]: let's re-use existing method, move existing ones to common folder if it's present in specific app

File: src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      return {
+        condition: valueField['condition'+valueField.rank] as ConditionField,
+        operator: valueField['operator'+valueField.rank] as Operator,
+        value: valueField['value'+valueField.rank],
+        join_by: valueField.join_by,
+        rank: valueField.rank
+      };
+    }
+
+    static constructSkipExportPayload(valueField: ExpenseFilterPayload, valueOption: any[]): ExpenseFilterPost {
```
Comment (ashwin1111) [inline src/app/core/models/business-central/business-central-configuration/business-central-advanced-settings.model.ts:40]: let's re-use existing method, move existing ones to common folder if it's present in specific app

File: src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts [hunk 5]
```diff
@@ -1,9 +1,58 @@
...
+  }
+
+  @CacheBuster({
+    cacheBusterNotifier: businessCentralAdvancedSettingGetCache$
+  })
+  postAdvancedSettings(advancedSettingsPayload: BusinessCentralAdvancedSettingsPost): Observable<BusinessCentralAdvancedSettingsGet> {
+    return this.apiService.post(`/workspaces/${this.workspaceService.getWorkspaceId()}/advanced_settings/`, advancedSettingsPayload);
+  }
+
+  getExpenseFields(): Observable<ConditionField[]> {
```
Comment (ashwin1111) [inline src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts:39]: where is this required?

File: src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts [hunk 6]
```diff
@@ -1,9 +1,58 @@
...
+    return this.apiService.get(`/workspaces/${this.workspaceService.getWorkspaceId()}/fyle/expense_fields/`, {});
+  }
+
+  getAdminEmail(): Observable<EmailOption[]> {
+    return this.apiService.get(`/workspaces/${this.workspaceService.getWorkspaceId()}/admin`, {});
+  }
+
+  getExpenseFilter(): Observable<ExpenseFilterResponse> {
```
Comment (ashwin1111) [inline src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts:47]: let's re-use existing method, move existing ones to common folder if it's present in specific app

File: src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts [hunk 7]
```diff
@@ -1,9 +1,58 @@
...
+    return this.apiService.get(`/workspaces/${this.workspaceService.getWorkspaceId()}/fyle/expense_filters/`, {});
+  }
+
+  postExpenseFilter(expenseFilter: ExpenseFilterPost): Observable<ExpenseFilter> {
```
Comment (ashwin1111) [inline src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts:51]: let's re-use existing method, move existing ones to common folder if it's present in specific app

File: src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts [hunk 8]
```diff
@@ -1,9 +1,58 @@
...
+    return this.apiService.post(`/workspaces/${this.workspaceService.getWorkspaceId()}/fyle/expense_filters/`, expenseFilter);
+  }
+
+  deleteExpenseFilter(rank: number): Observable<{}> {
```
Comment (ashwin1111) [inline src/app/core/services/business-central/business-central-configuration/business-central-advanced-settings.service.ts:55]: let's re-use existing method, move existing ones to common folder if it's present in specific app

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**